src/compiler/checker.ts

```diff
@@ -6703,8 +6790,11 @@ module ts {
                 }
                 else {
                     var exprType = checkExpression(<Expression>node.body);
-                    if (node.type) {
-                        checkTypeAssignableTo(exprType, getTypeFromTypeNode(node.type), node.body, /*headMessage*/ undefined);
+                    if (returnType) {
+                        if (isAsync) {
+                            exprType = getAwaitedType(exprType, /*fallbackType*/ exprType);
+                    }
+                        checkTypeAssignableTo(exprType, returnType, node.body, /*headMessage*/ undefined);

```

Comment:
+1

Also, doesn't getAwaitedType unwrap the promise? It seems like you are unwrapping the expression and checking if it's assignable to the return annotation. But I would think you want the opposite, namely get the expression so that it is wrapped in exactly one promise, and then check assignability.